src/bin/languageserver/mod.rs

```diff
@@ -36,6 +63,7 @@ pub struct SolangServer {
     importpaths: Vec<PathBuf>,
     importmaps: Vec<(String, PathBuf)>,
     files: Mutex<Files>,
+    definitions: Mutex<Definitions>,
```

Comment:
Here you have a single, global vector of definitions. That's not correct, I think. Say you have two files:
```solidity
contract X {
    struct S { int f1; }
    constructor(S s) {} 
}
```
and 
```solidity
contract X {
    struct S { int64 f1; }
    constructor(S s) {} 
}
```
Now you have two files that have definitions `DefinitionIndex::Struct(0)`. I think there should be a list of definitions per file.

Reply:
Hello @seanyoung ! 
In this case, won't the two contracts have different ids (i.e, they will be located at different indices of `Namespace::contracts`)? 
Also, this is the definition of the type `Definitions`. 
```rust
type Definitions = HashMap<DefinitionIndex, (PathBuf, Range)>;
```
It's a mapping from the `DefinitionIndex` (which will be unique for each "code object" in the namespace) to a tuple of filename where the "code object" is defined and its location within the file. 

So, we are in a way keeping track of the file information where something is defined, just in a different way. 
I chose to do it this way as the most frequent operation on `Definitions` is to look them up using their `DefinitionIndex` to get the file and location within the file where they are defined. So, I decided that grouping them by files(which is what we do with `hovers`, `references`) makes less sense for `definitions`.

Reply:
> In this case, won't the two contracts have different ids (i.e, they will be located at different indices of `Namespace::contracts`)?

Since they are different files, and do not import each other, they are separate `Namespace` structures.

Reply:
The `definitions` field is now part of `Files`. 
But this leads to another problem when the file is closed. This triggers the `did_close` handler function where we remove the entry corresponding to the closed file from `Files::cache`. 
But there can be live references in other files whose definitions are defined in the closed file. This is an issue. 
So, we either don't remove entries on file close, or we need to implement some kind of garbage collector that removes the entries only when we can confirm that there are no references pointing to the definitions in the closed file.

Reply:
You are right, this is a problem.

In fact if you open a file, then all the files this file imports are added to cache, but only removed if the files are opened and then closed. 

One solution is to only add the file which was requested in `fn parse_file` to the cache, and keep all its definitions in the same `FileCache` struct.

Reply:
Hello @seanyoung ! 
I have been thinking about this problem .
1. The path of the file where something is defined should be part of `DefinitionIndex` (which is not the case, as of now).  This is because the role of `DefinitionIndex` is to uniquely identify the location where something is defined and as it stands now, `DefinitionIndex` fails to do so when files belonging to multiple namespaces are opened. 
Making the filepath part of `DefinitionIndex` will help solve this issue. 
File path not being part of `DefinitionIndex` requires us to store the filepath in a tuple and makes things awkward down the line. 
2. If we make the above change, then we don't need to make `definitions` field part of `Files`. This will enable a "flat" lookup as opposed to going through `FileCache` first. This can help with the above issue that occurs when a file is closed. (https://github.com/hyperledger/solang/pull/1411#discussion_r1278318147)
By making the structs not be part of `Files`, we are free to handle the removal of definitions independently from that of `Files`. 
3. The whole filepath will have to be stored in the `DefinitionIndex` as of now. Using numbers to represent files will not work as the same number can point to two or more different files when multiple namespaces are involved. 
But I believe we should be able to handle this in the future with a layer of indirection that properly maps numbers to filepaths, despite the presence of multiple namespaces.

Reply:
Hi @chioni16
That makes a lot of sense, sounds good to me. It's fine to start the filepath in the `DefinitionIndex`. We can think of optimizing this later. I would expect there is a rust crate to help with this, but I have not been able to find one this morning.